elephant/spike_train_processing.py

```diff
@@ -0,0 +1,317 @@
+    [2 1 1 1 1]
+    >>> print(ep1.times)
+    [0.  3.5 4.5 5.5 7.5] ms
+    >>> print(ep1.durations)
+    [1.5 1.  1.  1.  1. ] ms
+
+    >>> # spread = 1, consecutive spikes
+    >>> ep2 = precise_complexity_intervals([st1, st2], sampling_rate, spread=1)
+    >>> print(ep2.array_annotations['complexity'].flatten())
+    [2 3 1]
+    >>> print(ep2.times)
+    [0.  3.5 7.5] ms
+    >>> print(ep2.durations)
+    [1.5 3.  1. ] ms
+
+    >>> # spread = 2, consecutive spikes and separated by 1 empty bin
+    >>> ep3 = precise_complexity_intervals([st1, st2], sampling_rate, spread=2)
+    >>> print(ep3.array_annotations['complexity'].flatten())
+    [2 4]
+    >>> print(ep3.times)
+    [0.  3.5] ms
+    >>> print(ep3.durations)
+    [1.5 5. ] ms
+
+    """
+    _check_consistency_of_spiketrainlist(spiketrains)
+
+    bin_size = 1 / sampling_rate
+
+    bst = conv.BinnedSpikeTrain(spiketrains,
+                                binsize=bin_size)
+
+    bincount = np.array(bst.to_sparse_array().sum(axis=0)).squeeze()
+
+    if spread == 0:
+        bin_indices = np.nonzero(bincount)[0]
+        complexities = bincount[bin_indices]
+        left_edges = bst.bin_edges[bin_indices]
+        right_edges = bst.bin_edges[bin_indices + 1]
+    else:
+        i = 0
+        complexities = []
+        left_edges = []
+        right_edges = []
+        while i < len(bincount):
+            current_bincount = bincount[i]
+            if current_bincount == 0:
+                i += 1
+            else:
+                last_window_sum = current_bincount
+                last_nonzero_index = 0
+                current_window = bincount[i:i + spread + 1]
+                window_sum = current_window.sum()
+                while window_sum > last_window_sum:
+                    last_nonzero_index = np.nonzero(current_window)[0][-1]
+                    current_window = bincount[i:
+                                              i + last_nonzero_index
+                                              + spread + 1]
+                    last_window_sum = window_sum
+                    window_sum = current_window.sum()
+                complexities.append(window_sum)
+                left_edges.append(
+                    bst.bin_edges[i].magnitude.item())
+                right_edges.append(
+                    bst.bin_edges[
+                        i + last_nonzero_index + 1
+                    ].magnitude.item())
+                i += last_nonzero_index + 1
+
+        # we dropped units above, because neither concatenate nor append works
+        # with arrays of quantities
+        left_edges *= bst.bin_edges.units
+        right_edges *= bst.bin_edges.units
+
+    # ensure that spikes are not on the bin edges
+    bin_shift = .5 / sampling_rate
+    left_edges -= bin_shift
+    right_edges -= bin_shift
+
+    # ensure that an epoch does not start before the minimum t_start
+    min_t_start = min([st.t_start for st in spiketrains])
```

Comment:
no need to pass `[` and `]`, `min()` works with iterators
so a simpler and valid version is `min(st.t_start for st in spiketrains)`